### DISCUSSION TITLE
creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#5] [Announcements]
# creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/

In this post I describe how to create parameterised tests using xUnit's [Theory], [InlineData], [ClassData], and [MemberData] attributes.

https://andrewlock.net/creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/

**Comment by VosWouter87** [comment]
It's hard to identify between the different testcases. I'm using InlineData and now have to hover over the testresult to see which one failed. Is there a way to include that data in the message of a failed test?

**Reply by FerryMan-Coder** [reply]
This is quite late, but in case someone else wonders: You could use ITestOutputHelper (part of xUnit) to write messages: Call  _testOutputHelper.WriteLine(), for instance. You will then be able to write these messages to the same output as where you see the rest of your test results.

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 08 2017, 09:18, in reply to [Paulo Morgado](https://disqus.com/by/paulo_morgado/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/5#discussioncomment-6043057)</em>

---

<p>Interesting, thanks for pointing those out, I'll look into them :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 16 2019, 06:59, in reply to [John Moser](https://disqus.com/by/disqus_fr1SeSOWUC/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/5#discussioncomment-6043074)</em>

---

<p>parameters is a params array, so you would use it like this:</p><p><code>[MemberData(nameof("GetData"), 3, "string", 1.0m)] </code></p><p>You can see the source code here: <a href="https://github.com/xunit/xunit/blob/master/src/xunit.core/MemberDataAttribute.cs" rel="nofollow noopener" title="https://github.com/xunit/xunit/blob/master/src/xunit.core/MemberDataAttribute.cs">https://github.com/xunit/xunit/blob/master/src/xunit.core/MemberDataAttribute.cs</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 15 2019, 10:29, in reply to [Viacheslav Fesenko](https://disqus.com/by/slavasergeevich/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/5#discussioncomment-6043080)</em>

---

<p>Good point, thanks!</p>

[a comment by an automated account is not shown]

**Comment by carbonell** [comment]
This article is so good that I still keep getting it from coworkers that are new to unit testing. Have you consider updating it to include [TheoryData] to declare strongly  typed test parameters on xunit tests?

**Reply by andrewlock** [reply]
Haha, thanks - I have a separate post about that here: https://andrewlock.net/creating-strongly-typed-xunit-theory-test-data-with-theorydata/ 😊

**Comment by elibroftw** [comment]
Can you update the article to suggest MemberData before ClassData to save on LOC? Also in .NET 9, you can simplify the initialization like this:

```cs
public static IEnumerable<object[]> YearTestData =>
        [
            [2010],
            [2019],
            [DateTime.Today.Year - 1],
        ];
```
